README
=============
Yowie library
=============

This Python/Django library provides utilities for safe reading large
file-like objects.


-----------------------------
yowie.open_url(url, **kwargs)
-----------------------------

This function open url and return file object. Url attribute can be local
file path or full url path. Allowed protocols are local file path, file,
http and ftp. kwargs are additional attributes according to protocol -
'mode' for local path and file protocol, 'proxy', 'data' and 'timeout'
(Python >= 2.6) for http and ftp protocols.

Examples:

open_url('/home/praetorian/secret.txt')
open_url('file:///home/praetorian/secret.txt', mode='r')
open_url('http://domain.tld/secret.txt', proxy='172:16:1:100:8000')
open_url('ftp://domain.tld/secret.txt')

----------------------------------------------------
class yowie.LimitedFile(file_descriptor, limit=None)
----------------------------------------------------

File-like object wrapper which raise LimitedFileSizeOverflow exception
if file content length is greater then limit. Default limit is
settings.LIMITEDFILE_MAX_SIZE bytes or 10MB, if setting doesn't exist.
Override read([size]), readline([size]) and readlines([sizehint]) methods.
Sum of data, which has been read, is stored in the total attribute.

-----------------------------------------------------
yowie.LimitedFile.open_url(url, limit=None, **kwargs)
-----------------------------------------------------

Static method, shortcut for:

fd = open_url('my_big_file.txt')
lf = LimitedFile(fd, 1024*1024)

or

lf = LimitedFile.open_url('my_big_file.txt', 1024*1024)

---------------------------------------
exception yowie.LimitedFileSizeOverflow
---------------------------------------

LimitedFile class raise this exception if file content length is greater
then limit.


Example:
--------

>>> from yowie import LimitedFile, open_url

>>> fd = open_url('my_big_file.txt')
>>> lf = LimitedFile(fd, 1024*1024)

>>> content = ''
>>> content += lf.read(1024*512)
>>> content += lf.read(1024*512)
>>> content += lf.read(1024*512)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib/python2.7/site-packages/yowie/fileutils.py", line 55, in read
    raise LimitedFileSizeOverflow('File content is too long')
yowie.fileutils.LimitedFileSizeOverflow: File content is too long

>>> lf.total
1048577


debian/changelog
centrum-python-yowie (0.3-1) unstable; urgency=low

  * Build from f886117

 -- Jan Seifert <jan.seifert@centrumholdings.com>  Wed, 26 Jan 2011 08:28:41 +0100

python-yowie (0.3) unstable; urgency=low

  * Build from fd1584a

 -- Jan Seifert <jan.seifert@centrumholdings.com>  Tue, 25 Jan 2011 16:39:45 +0100

python-yowie (0.1) unstable; urgency=low

  * initial build

 -- CH content team <pg-content-dev@chconf.com>  Tue, 11 Jan 2011 17:38:23 +0100



tests/test_project/settings.py
from os import path

DEBUG = False
TEMPLATE_DEBUG = DEBUG

DATABASE_ENGINE = 'sqlite3'
DATABASE_NAME = ':memory:'

ROOT_URLCONF = 'test_project.urls'

TEMPLATE_DIRS = (
    path.join(path.dirname(__file__), "templates"),
)

INSTALLED_APPS = (
    'test_fileutils',
)

EXAMPLE_RSS_FEED = path.join(path.dirname(__file__), "data", "example.rss")


yowie/fileutils.py
from urllib2 import build_opener, ProxyHandler, urlparse, URLError

from django.conf import settings
from django.core.files.utils import FileProxyMixin

__all__ = [
    "open_url",
    "LimitedFileSizeOverflow",
    "LimitedFile"
]

def open_url(url, **kwargs):
    """
    open_url(url, **kwargs) - open url and return file descriptor

    url - local file path or full url path. Allowed protocols are local file
    path, file, http and ftp

    kwargs - additional attributes according to protocol, 'mode' for local
    path and file protocol, 'proxy', 'data' and 'timeout' (Python >= 2.6)
    for http and ftp protocols

    Examples:

    open_url('/home/praetorian/secret.txt')
    open_url('file:///home/praetorian/secret.txt', mode='r')
    open_url('http://domain.tld/secret.txt', proxy='172:16:1:100:8000')
    open_url('ftp://domain.tld/secret.txt')
    """
    bits = urlparse.urlsplit(url)
    attrs = kwargs

    if bits.scheme in ('', 'file'):
        url = bits.netloc + bits.path
        opener = open
    elif bits.scheme in ('http', 'ftp'):
        handlers = []
        if 'proxy' in attrs:
            handlers.append(ProxyHandler({bits.scheme: attrs.pop('proxy')}))

        url =  bits.geturl()
        opener = build_opener(*handlers).open
    else:
        raise URLError("Unsupported protocol '%s'" % bits.scheme)

    return opener(url, **attrs)


class LimitedFileSizeOverflow(Exception):
    pass

class LimitedFile(FileProxyMixin):
    """
    File-like object wrapper which raise LimitedFileSizeOverflow
    exception if file content length is greater then limit. Default limit is
    settings.LIMITEDFILE_MAX_SIZE bytes or 10MB, if setting doesn't exist.

    LimitedFile(file_descriptor, limit=None)
    """

    _limit = getattr(settings, 'LIMITEDFILE_MAX_SIZE', 10*1024*1024)

    def __init__(self, file, limit=None):
        self.file = file
        self.limit = limit or self._limit
        self.total = 0

    def _normalize_size(self, size):
        if not size or size < 0 or size > self.limit:
            size = self.limit + 1

        rest = self.limit - self.total
        if size > rest:
            size = rest + 1

        return size

    @staticmethod
    def open_url(url, limit=None, **kwargs):
        """
        open_url(url [, limit [, **kwargs]]) - staticmethod, open url and
        return file descriptor

        url - url to opening, can be local file path, file://...,
        http://... and ftp://...

        limit - raise LimitedFileSizeOverflow exception if file content
        length is greater then limit

        kwargs - additional attributes according to protocol, 'mode' for
        local path and file protocol, 'proxy', 'data' and
        'timeout' (Python >= 2.6) for http and ftp protocols.

        Examples:

        LimitedFile.open_url('/home/praetorian/secret.txt')
        LimitedFile.open_url('file:///home/praetorian/secret.txt', mode='r')
        LimitedFile.open_url('http://domain.tld/secret.txt',
                             limit=1024*1024, proxy='172:16:1:100:8000')
        LimitedFile.open_url('ftp://domain.tld/secret.txt')
        """
        return LimitedFile(open_url(url, **kwargs), limit)

    def read(self, size=None):
        " read(self [, size]) "
        content = super(LimitedFile, self).read(self._normalize_size(size))
        self.total += len(content)

        if self.total > self.limit:
            raise LimitedFileSizeOverflow('File content is too long')

        return content

    def readline(self, size=None):
        " readline(self [, size]) "
        content = super(LimitedFile, self).readline(self._normalize_size(size))
        self.total += len(content)

        if self.total > self.limit:
            raise LimitedFileSizeOverflow('File content is too long')

        return content

    def readlines(self, sizehint=None):
        " readlines(self [, sizehint]) "
        content = self.read(self._normalize_size(sizehint))

        lines = []
        while content:
            index = content.find('\n')

            if index >= 0:
                lines.append(content[0:index + 1])
                content = content[index + 1:]
            else:
                lines.append(content)
                break

        return lines

    def next(self):
        " next(self) "
        line = self.readline()
        if not line:
            raise StopIteration()
        return line
